rfcs/0007-scoring-system-and-network-security/0007-scoring-system-and-network-security.md

```diff
@@ -0,0 +1,303 @@
+    check_sync_stale_at = now + check_sync_stale_interval
+  end
+end
+```
+
+CKB network will to continually try to connect to extra outbound peers when `try_new_outbound_peer` is `true`, and check outbound peers number every few minutes and trying to evict useless extra peers to prevent we have too many connections.
+
+``` ruby
+# eviction logic
+def evict_extra_outbound_peers
+  connected_outbound_peers = connected_peers.select{|peer| peer.outbound? && !peer.feeler? }
+  if connected_outbound_peers.length <= max_outbound
+    return
+  end
+  now = Time.now
+  # here use last_block_anoncement_at to evict peers, we assume the oldest one is useless for us
+  evict_target = connected_outbound_peers.sort_by do |peer|
+    peer.last_block_announcement_at
+  end.first
+  if evict_target
+    if now - evict_target.last_connected_at > MINIMUM_CONNECT_TIME && !is_downloading?(evict_target)
+      disconnect_peer(evict_target)
+      # prevent connect to too many peers
+      set_try_new_outbound_peer(false)
+    end
+  end
+end
+```
+
+### The process of accept inbound peers
+
+In Bitcoin, a node will try to evict connected inbound peers if the number of connected inbound peers reach `max_inbound` and another new inbound connection detected. (check [Bitcoin source code][1] for detail)
+
+This eviction behavior is intended to keep high-quality peer and evict low-quality peer.
+
+CKB should implement the eviction as the following steps:
+
+1. group connected inbound peers by `network group` field
+2. find the group which contains most peers
+3. find the lowest score in the group, evict the lowest scored peer if new inbound peer's score is higher than this; otherwise, we disconnect the new peer
+
+### Feeler Connection
+
+Feeler Connection is intended to：
+
+* Test a peer is connectable or not
+* Discovery more address to fill the peer store
+
+Node will start a feeler connection every few minutes after outbound peers reach `max_outbound` limit.
+
+1. Randomly choice peer info from peer store which score should higher than `TRY_SCORE`
+2. Connect to peer
+3. Run node discovery protocol with the peer
+4. Disconnect
+
+Feeler peers should be assumed to be disconnected soon.
+
+### delete peer info from peer store
+
+Required parameters:
+
+* `PEER_STORE_LIMIT` - max number of peer info in peer store
+* `PEER_NOT_SEEN_TIMEOUT` - used to protect peers which we recently connected, we only delete peer info which `last_connected_to` is older than this value. 
+
+When the number of peer info reach `PEER_STORE_LIMIT`:
+
+1. group all peer infos in peer store by `network group` field
+2. find the group which contains most peer infos
+3. find peers we have not seen recently from this group: `peer.last_connected_at < Time.now - PEER_NOT_SEEN_TIMEOUT`
+4. find lowest scored peer info as `candidate_peer_info`
+5. if `candidate_peer_info.score < new_peer_info.score` than we delete `candidate_peer_info` and add `new_peer_info`, otherwise we do not accept `new_peer_info`
```

Comment:
```suggestion
5. if `candidate_peer_info.score < new_peer_info.score` then we delete `candidate_peer_info` and add `new_peer_info`, otherwise we do not accept `new_peer_info`
```